### ISSUE TITLE
Improve download support

### ISSUE THREAD

**Issue by thornbill** [jellyfin/jellyfin-ios#372]
The following improvements are needed to the initial download implementation:

* [x] Delete transcodes on server post-download &mdash; #408
* [x] Fetch item info
  * Starting in 10.10 web should pass the item details to the download function &mdash; https://github.com/jellyfin/jellyfin-web/pull/6214
* [x] Verify item is a video before attempting to download &mdash; #637
* [x] Add messaging on empty download screen &mdash; #636 
* [x] Support direct download of compatible media &mdash; #708
* [x]  Limit the number of concurrent downloads &mdash; #711
* [x] Investigate support for opening downloads in other apps &mdash; #718
* [ ] Improve UI of download listing
* [ ] Save playback progress
* [ ] Add quality options for downloads
* [ ] Display download file sizes
* [ ] Display available space on the device
* [ ] Support subtitles for downloads
* [ ] Allow cancelling in progress or pending downloads
* [ ] Allow retrying failed downloads
* [ ] Make downloads pausable/resumable (probably impossible without server changes)

**Comment by dm3ch** [comment]
Hi! Is there any progress with publically releasing this feature?

**Comment by thornbill** [comment]
No there is nothing new to report... I've been slacking on working on this app. 😅

**Comment by dm3ch** [comment]
Is there anything that newby could try to help with?

**Comment by NoSadBeHappy** [comment]
Once this is finished, will it be integrated in the app to get to the downloaded content? Also will it be supported for music?

**Comment by anthonylavado** [comment]
@NoSadBeHappy Yes, it would be integrated in the app. It would just be video for now. For music, apps like Finamp do support offline downloads/playback.

**Comment by NoSadBeHappy** [comment]
> @NoSadBeHappy Yes, it would be integrated in the app. It would just be video for now. For music, apps like Finamp do support offline downloads/playback.

Thanks for the quick response! Sounds great. Yeah i knew about finamp, but i don't like the interface as much as the official app, and it also doesen't support artist sorting in offline mode, which is how I ususally like to choose my music.

**Comment by AlexandruNegura** [comment]
Waiting for this one as well. It would make Plex Pass obsolete :)

**Comment by hgghyxo** [comment]
Is it possible to introduce a cache to disk system that supports future downloading capabilities ? 
Currently it feels like every time I reopen the player ( mostly due to very bad network on a train ) it has to download chunks again, even if it was already downloaded. 
I do a bit of JS coding myself, I might try to look into this later myself, but would like to align with your projects and plan. 
Thanks

**Comment by pawisoon** [comment]
Hey! I wanted to ask about the plans for this feature. I see its marked as "todo" in Jellyfin iOS Roadmap since may 2023. Is there any plans to make it happen in 2025?
It would be an absolute banger. I noticed Android has this functionality but iOS is lacking behind :(

Thank you for your work and looking forward upcoming awesome features :)

**Comment by thornbill** [comment]
You can read the latest update in the [1.6 release blog post](https://jellyfin.org/posts/ios-v1.6.0).

**Comment by pawisoon** [comment]
> You can read the latest update in the [1.6 release blog post](https://jellyfin.org/posts/ios-v1.6.0).

Thank you! I didnt see this post. Im super happy to see this announcement addressing the download feature! Cheers!

**Comment by sjorge** [comment]
Going to comment here as I am not sure it warrants a separate issue.

The nice thing about the downloads showing ip in files under my
iPhone is that other apps can be used to play them, e.g. if the native player does not support the subtitle format.

The big downside is that they will be included in the iCloud backups!

Mine luckily failed as it would have been over 50Gb over cellular!

I found no way to exclude the Jellyfin-owned files from the backup.

It would be great if we somehow could. I know of other apps where I can toggle the inclusion, or they are just not included. It would be good if we could do the same, but that might mean storing the data in some other way 😫

Edit: For some reason the testflight build shows up as org.jellyfin<guid> not as jellyfin like the normal app build. I can exclude it but for a manual backup it now always shows an extra popup, let's see if the nightly auto backup works or not.

**Comment by sjorge** [comment]
Another minor issue:

When downloading a whole show that has 2 files for the same episode, both get downloaded. The 2nd one replaces the first one but it will have 2 entries on the downloads tap. Deleting either one leaves the other one as broken.

Not sure what should happen here, probably not download both?

**Comment by JordanMPDS** [comment]
> Going to comment here as I am not sure it warrants a separate issue.
> 
> The nice thing about the downloads showing ip in files under my iPhone is that other apps can be used to play them, e.g. if the native player does not support the subtitle format.
> 
> The big downside is that they will be included in the iCloud backups!
> 
> Mine luckily failed as it would have been over 50Gb over cellular!
> 
> I found no way to exclude the Jellyfin-owned files from the backup.
> 
> It would be great if we somehow could. I know of other apps where I can toggle the inclusion, or they are just not included. It would be good if we could do the same, but that might mean storing the data in some other way 😫
> 
> Edit: For some reason the testflight build shows up as org.jellyfin not as jellyfin like the normal app build. I can exclude it but for a manual backup it now always shows an extra popup, let's see if the nightly auto backup works or not.

According to apple support you can deselect an app from being included in the backup [link here](https://support.apple.com/en-qa/108922) under the section "Reduce the size of your iCloud Backup

> Choose which apps to back up on your iPhone, iPad, or Apple Vision Pro
> 
> Open the Settings app, tap your name, then tap iCloud.
> 
> Tap Storage or tap Manage Account Storage, then tap Backups.
> 
> Tap the name of the device that you're using.
> 
> Turn off any apps that you don't want to back up.
> 
> Tap Turn Off to confirm.
> 
> When you confirm that you want to Turn Off, it turns off iCloud Backup for that app and removes all of its information from iCloud.

**Comment by AdriaanBoshoff** [comment]
Please also allow cancellations of downloads. At the moment you have to reset the app to stop downloads since theres no way to cancel them.

**Comment by rickvian** [comment]
does anyone working on improving UI download right now?

**Comment by Mumbolio85** [comment]
I just made a showcase for what I think the downloads page could look like if the UI was overhauled. I put it as a PR on my own branch of Jellyfin-ios because I am aware that there is a process for implimenting new UIs. Here is the link to the showcase https://github.com/Mumbolio85/jellyfin-ios/pull/1. 

The redesign makes the downloads page a tile based grid, much like the movies library but without the cover poster images. The items are grouped by media type, and episodes/tracks in a show/album are further grouped together for easier finding of those items. The page can be sorted by date added or name, and shows/albums are always sorted by item order. New items which are part of a show that already has items downloaded will show the item outside the group while downloading, allowing the user to keep track of what has yet to finish downloading.

Here are some screenshots of what the redesign looks like:

<img width="195" height="422" alt="IMG_3886" src="https://github.com/user-attachments/assets/4f274c3f-ed9f-4e3c-b751-98637ba5d3be" /> 

<img width="195" height="422" alt="IMG_3888" src="https://github.com/user-attachments/assets/6939d1ea-1115-40b1-b4c7-f353e7522910" />

<img width="195" height="422" alt="IMG_3891" src="https://github.com/user-attachments/assets/a93bc805-32ca-4122-82f8-5da6743234e7" />

<img width="195" height="422" alt="IMG_3887" src="https://github.com/user-attachments/assets/56c715db-51cc-4271-8c8f-d912ffcfee65" />


If people like what I came up with, I would be happy to start puting it through whatever process is needed when testing new UIs.

**Comment by cncptpr** [comment]
I don't know if it is planned, but I did not see it listed and it doesn't work right now: playing the media in the actual app.

The video is currently just downloaded as a file, and must be played with an external player. The apple native one is not good for long videos, and doesn't support too many formats. Also progress is obviously not synced.

It would be cool to just watch it in the app.
